jdk/src/share/classes/java/security/Provider.java

```diff
@@ -920,7 +928,7 @@ private void removeInvalidServices(Map<ServiceKey,Service> map) {
         for (Iterator<Map.Entry<ServiceKey, Service>> t =
                 map.entrySet().iterator(); t.hasNext(); ) {
             Service s = t.next().getValue();
-            if (s.isValid() == false) {
+            if ((s.isValid() == false) || !RestrictedSecurity.isServiceAllowed(s)) {
```

Comment:
In the method “ensureLegacyParsed()”, it will first call the “parseLegacyPut()” method to register the legacy services. And then, call the “removeInvalidServices()” method to remove the invalid and also the restricted security not allowed services. So, the threes checks in the “parseLegacyPut()” method are not needed anymore. So, I removed them.